File: src/year2024/day22/day22.rs
//! [Day 22: Monkey Market](https://adventofcode.com/2024/day/22)

use rayon::prelude::*;

const fn next_secret(mut secret: i64) -> i64 {
    /*
    let secret = (secret ^ (secret * 64)) % 16_777_216;
    let secret = (secret ^ (secret / 32)) % 16_777_216;
    (secret ^ (secret * 2048)) % 16_777_216
    */
    secret ^= secret << 6;
    secret &= 0xFF_FFFF;
    secret ^= secret >> 5;
    secret ^= secret << 11;
    secret &= 0xFF_FFFF;
    secret
}

struct Puzzle {
    initial_secrets: Vec<i64>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        Self {
            initial_secrets: data.lines().map_while(|s| s.parse::<i64>().ok()).collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> i64 {
        self.initial_secrets
            .par_iter()
            .map(|&initial_secret| (0..2000).fold(initial_secret, |secret, _| next_secret(secret)))
            .sum()
    }

    /// Solve part two.
    fn part2(&self) -> i64 {
        const RANGE: usize = 19;
        const SIZE: usize = RANGE * RANGE * RANGE * RANGE;

        let total_bananas: Vec<i32> = self
            .initial_secrets
            .par_iter()
            .fold(
                || vec![0; SIZE],
                |mut acc, &initial_secret| {
                    let mut seen = vec![false; SIZE];
                    let mut secret = initial_secret;
                    let mut idx = 0;

                    for i in 0..2000 {
                        let next = next_secret(secret);
                        let p_prev = (secret % 10) as i8;
                        let p_curr = (next % 10) as i8;
                        let diff = p_prev.abs_diff(p_curr + 9) as usize;

                        idx = (idx % 6859) * 19 + diff;

                        if i >= 3 && !seen[idx] {
                            acc[idx] += i32::from(p_curr);
                            seen[idx] = true;
                        }
                        secret = next;
                    }
                    acc
                },
            )
            .reduce(
                || vec![0; SIZE],
                |mut a, b| {
                    for (x, y) in a.iter_mut().zip(b.iter()) {
                        *x += *y;
                    }
                    a
                },
            );

        i64::from(*total_bananas.iter().max().unwrap_or(&0))
    }
}

/// # Panics
#[must_use]
pub fn solve(data: &str) -> (i64, i64) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const TEST_INPUT: &str = include_str!("test.txt");

    #[test]
    fn test_part1() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part1(), 37327623);
    }

    #[test]
    fn test_part2() {
        let puzzle = Puzzle::new(TEST_INPUT);
        assert_eq!(puzzle.part2(), 23 + 1);
    }
}

File: src/year2024/day22/day22.go
// [Day 22: Monkey Market](https://adventofcode.com/2024/day/22)

package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"time"
)

// nextSecret computes the next secret value using the same small PRNG
// transformations as the Rust implementation. All operations are kept
// in 24-bit space (mod 2^24).
func nextSecret(secret int64) int64 {
	secret = (secret ^ (secret * 64)) % 16_777_216
	secret = (secret ^ (secret / 32)) % 16_777_216
	secret = (secret ^ (secret * 2048)) % 16_777_216
	return secret
}

// Puzzle holds the parsed initial secrets.
type Puzzle struct {
	initialSecrets []int64
}

// NewPuzzle parses the input lines and collects initial secrets until a
// non-integer line is encountered (behaves like Rust's map_while).
func NewPuzzle(data string) *Puzzle {
	var secs []int64
	for _, line := range strings.Split(strings.TrimSpace(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		v, err := strconv.ParseInt(line, 10, 64)
		if err != nil {
			break
		}
		secs = append(secs, v)
	}
	return &Puzzle{initialSecrets: secs}
}

// Part1 applies nextSecret 2000 times to each initial secret and returns
// the sum of final secrets.
func (p *Puzzle) Part1() int64 {
	var sum int64
	for _, initial := range p.initialSecrets {
		secret := initial
		for i := 0; i < 2000; i++ {
			secret = nextSecret(secret)
		}
		sum += secret
	}
	return sum
}

// Part2 analyses the price-digit sequences and accumulates banana values
// grouped by the sequence of four successive differences.
func (p *Puzzle) Part2() int64 {
	bananas := make(map[[4]int64]int64)

	for _, initial := range p.initialSecrets {
		prices := make([]int64, 0, 2001)

		secret := initial
		prices = append(prices, secret%10)
		for i := 0; i < 2000; i++ {
			secret = nextSecret(secret)
			prices = append(prices, secret%10)
		}

		seen := make(map[[4]int64]struct{})
		for i := 0; i+4 < len(prices); i++ {
			p0 := prices[i : i+5]
			seq := [4]int64{p0[1] - p0[0], p0[2] - p0[1], p0[3] - p0[2], p0[4] - p0[3]}
			if _, ok := seen[seq]; !ok {
				seen[seq] = struct{}{}
				bananas[seq] += p0[4]
			}
		}
	}

	var best int64
	for _, v := range bananas {
		if v > best {
			best = v
		}
	}

	return best
}

// solve is the package-level solver used by main and tests.
func solve(data []byte) (int64, int64) {
	p := NewPuzzle(string(data))
	return p.Part1(), p.Part2()
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		r1, r2 := solve(data)
		duration := time.Since(start)

		fmt.Println(r1)
		fmt.Println(r2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		r1, r2 := solve(data)
		fmt.Println(r1)
		fmt.Println(r2)
	}
}
